### ISSUE TITLE
Misleading error message for deeply nested object literal with union type and unknown property

### ISSUE THREAD

**Issue by duanyao** [Microsoft/TypeScript#8079]
**TypeScript Version:** 

1.8.10

**Code**

```ts
interface TNode {
	Kids?: (TNode | string)[];
}

function foo(node: TNode) { }

foo({
	Kids: [ { Kids: [ { c: 1 } ]} ]
});
```

**Expected behavior:**
The error message should contains something like "Object literal may only specify known properties, and 'c' does not exist in type 'TNode | string' "

**Actual behavior:** 
The error message is
```
error TS2345: Argument of type '{ Kids: { Kids: { c: number; }[]; }[]; }' is not assignable to parameter of type 'TNode'.
  Types of property 'Kids' are incompatible.
    Type '{ Kids: { c: number; }[]; }[]' is not assignable to type '(TNode | string)[]'.
      Type '{ Kids: { c: number; }[]; }' is not assignable to type 'TNode | string'.
        Type '{ Kids: { c: number; }[]; }' is not assignable to type 'string'.
```

The misleading part is "Type '{ Kids: { c: number; }[]; }' is not assignable to type 'string' ".

If I change `TNode | string` to `TNode | Element`, the error message becomes
```
error TS2345: Argument of type '{ Kids: { Kids: { c: number; }[]; }[]; }' is not assignable to parameter of type 'TNode'.
  Types of property 'Kids' are incompatible.
    Type '{ Kids: { c: number; }[]; }[]' is not assignable to type '(TNode | Element)[]'.
      Type '{ Kids: { c: number; }[]; }' is not assignable to type 'TNode | Element'.
        Type '{ Kids: { c: number; }[]; }' is not assignable to type 'Element'.
          Property 'classList' is missing in type '{ Kids: { c: number; }[]; }'.
```
Which is also misleading.

Strangely, If I reverse the order of the union types (i,e. `string | TNode` and `Element | TNode`), the error messages become expected.

**Comment by mhegazy** [comment]
It is as you have explained. the compiler has no way of knowing which error message is "better". and frankly it does not even try. if you have a proposal for reporting better error messages we would be happy to look into it.

**Comment by duanyao** [comment]
As I mentioned above, the better error message is already there if I reverse the order of the union type:
```ts
interface TNode {
    Kids?: (string | TNode)[]; // note the order!
}

function foo(node: TNode) { }

foo({
    Kids: [ { Kids: [ { c: 1 } ]} ]
});
```
The error message becomes expected:
```
error TS2345: Argument of type '{ Kids: { Kids: { c: number; }[]; }[]; }' is not assignable to parameter of type 'TNode'.
  Types of property 'Kids' are incompatible.
    Type '{ Kids: { c: number; }[]; }[]' is not assignable to type '(string | TNode)[]'.
      Type '{ Kids: { c: number; }[]; }' is not assignable to type 'string | TNode'.
        Type '{ Kids: { c: number; }[]; }' is not assignable to type 'TNode'.
          Types of property 'Kids' are incompatible.
            Type '{ c: number; }[]' is not assignable to type '(string | TNode)[]'.
              Type '{ c: number; }' is not assignable to type 'string | TNode'.
                Object literal may only specify known properties, and 'c' does not exist in type 'string | TNode'.
```
It seems a bug that `string | TNode` is not equivlant to `TNode | string`.

**Comment by RyanCavanaugh** [comment]
Which error is "better" is totally arbitrary (it's easy to construct scenarios where a longer error message is better than a shorter one, and vice versa). See #7785 as well.

**Comment by duanyao** [comment]
So how about show both versions of error messages? If the compiler is not smart enough to judge which message is better, let developers do it.

**Comment by mhegazy** [comment]
that will be much more text, (consider a union with 4 or 5 constituents, which are common) most of the time they are not much different, and users will complain that the useful information is buried under tons of useless ones..

**Comment by duanyao** [comment]
Sure, but I think it is better than no useful information at all.
